elements/lisk-p2p/test/unit/p2p_request.ts

```diff
@@ -0,0 +1,164 @@
+			expect(request)
+				.to.have.property('wasResponseSent')
+				.which.equals(false);
+		});
+
+		it('should increment the productivity.requestCounter by 1', async () => {
+			expect(requestOptions.productivity.requestCounter).to.equal(1);
+		});
+
+		it('should increment the productivity.requestCounter by 2 if a second P2PRequest instance is created with the same productivity tracker', async () => {
+			// P2PRequest instance can mutate the productivity tracker object.
+			new P2PRequest(requestOptions, respondCallback);
+			expect(requestOptions.productivity.requestCounter).to.equal(2);
+		});
+
+		it('should initiate productivity.responseCounter with the value specified in the constructor', async () => {
+			expect(requestOptions.productivity.responseCounter).to.equal(0);
+		});
+	});
+
+	describe('#end', () => {
+		let timeBeforeLastResponse: number;
+		beforeEach(() => {
+			timeBeforeLastResponse = Date.now();
+			request.end('hello');
+		});
+
+		it('should send data back to callback in correct format', async () => {
+			expect(respondCallback).to.be.calledOnceWith(undefined, {
+				data: 'hello',
+			});
+		});
+
+		it('should increment the productivity.responseCounter by 1', async () => {
+			expect(requestOptions.productivity.responseCounter).to.equal(1);
+		});
+
+		it('should have a productivity.responseRate of 1; this indicates a success rate of 100%', async () => {
+			expect(requestOptions.productivity.responseRate).to.equal(1);
+		});
+
+		it('should increment the productivity.responseCounter by 2 if a second P2PRequest instance is ended', async () => {
+			// P2PRequest instance can mutate the productivity tracker object.
+			let secondP2PRequest = new P2PRequest(requestOptions, respondCallback);
+			secondP2PRequest.end('world');
+
+			expect(requestOptions.productivity.responseCounter).to.equal(2);
+		});
+
+		it('should have a productivity.lastResponded which represents the time of the last successful response in milliseconds', async () => {
+			expect(requestOptions.productivity.lastResponded).to.gte(
+				timeBeforeLastResponse,
+			);
+		});
+
+		it('should set wasResponseSent property to true', async () => {
+			expect(request)
+				.to.have.property('wasResponseSent')
+				.which.equals(true);
+		});
+	});
+
+	describe('#when error', () => {
+		let err = new Error('Custom error');
+		err.name = 'CustomError';
+
+		beforeEach(() => {
+			request.error(err);
+		});
+
+		it('should send data back to callback in correct format', async () => {
+			expect(respondCallback).to.be.calledOnceWith(err);
+		});
+
+		it('should not increment the productivity.responseCounter', async () => {
+			expect(requestOptions.productivity.responseCounter).to.equal(0);
+		});
+
+		it('should have a productivity.responseRate of 0; this indicates a success rate of 0%', async () => {
+			expect(requestOptions.productivity.responseRate).to.equal(0);
+		});
+	});
+
+	describe('#when mixed success and error', () => {
```

Comment:
This is particular case of `#error` method. It should be inside its describe as we have discussed.